### ISSUE TITLE
JavaScript removed after deployment (also in v1.13.3)

### ISSUE THREAD

**Issue by kristoffer-zliide** [microsoft/vscode-azurefunctions#3978]
We have the following directory structure:
```
build/
  azure/
    host.json
    my-function/
      function.json
      index.js
```
and this in `.vscode/settings.json`:
```
  "azureFunctions.projectLanguage": "JavaScript",
  "azureFunctions.deploySubpath": "build/azure/",
```
Deploying with v.1.13.2 obviously destroyed our deployment, due to #3969 and v1.13.3 seems to be at least looking in the right directory. However, critically, it now seemingly ignores the `index.js` file. Output from Azure Functions in VSCode:
```
10:38:02 AM test-my-function: Starting deployment...
10:38:02 AM test-my-function: Creating zip package...
10:38:02 AM test-my-function: Adding 3 files to zip package...
10:38:02 AM test-my-function: c:\Users\...\Source\my-function\build\azure\local.settings.json
10:38:02 AM test-my-function: c:\Users\...\Source\my-function\build\azure\host.json
10:38:02 AM test-my-function: c:\Users\...\Source\my-function\build\azure\my-function\function.json
10:38:02 AM test-my-function: Zip package size: 981 B
10:38:06 AM test-my-function: Updating submodules.
10:38:06 AM test-my-function: Preparing deployment for commit id '86...0f'.
10:38:06 AM test-my-function: Skipping build. Project type: Run-From-Zip
10:38:07 AM test-my-function: Skipping post build. Project type: Run-From-Zip
10:38:07 AM test-my-function: Triggering recycle (preview mode disabled).
10:38:07 AM test-my-function: Deployment successful.
10:38:21 AM test-my-function: Syncing triggers...
10:38:23 AM test-my-function: Querying triggers...
10:38:25 AM test-my-function: No HTTP triggers found.
```

981B zip file doesn't look good, and the "Code + Test" tab in the Portal confirms that only the function.json file is in the function folder.

**Comment by nturinski** [comment]
Do you happen to have a `.gitignore` file in your directory? I can't really think of any reason why your `function.json` file is getting added to the zip file, but not the `index.js` if they are in the same folder.

**Comment by kristoffer-zliide** [comment]
Of course, I have a `.gitignore` ☺️ Why do you ask?

**Comment by kristoffer-zliide** [comment]
We don't have a `.funcignore` (since we have the build output in a separate folder), if that's what you mean. Did you reintroduce [the bug](https://github.com/microsoft/vscode-azurefunctions/issues/3817#issuecomment-1696236551) where `.gitignore` is confused for `.funcignore`?

**Comment by nturinski** [comment]
Oops, haha yeah, you're right, I meant `.funcignore`. As far as I know, that bug has not been reintroduced 😅

The relevant code for acquiring files to zip is right here: https://github.com/microsoft/vscode-azuretools/blob/main/appservice/src/deploy/runWithZipStream.ts#L136-L149 (ignore the name of the function, it is using the `.funcignore` path for this). Really not sure why it's totally ignoring the `index.js`. I'll try to see if I can reproduce this.

Just to confirm, this was working prior to v1.13.2?

**Comment by kristoffer-zliide** [comment]
Yeah, was also looking at [this](https://github.com/microsoft/vscode-azuretools/pull/1616/commits/a36b21288ebfbafa02a3378bcf87889bcd0476cb#diff-4687132758bf0251f89fb7089a9b97ac1a8e7511b49f9310454da7be91874edfR120), on the surface it looks OK... though the test coverage could probably be improved 😅 

Yes, it was working fine up to and including v.1.13.1. Also, to be clear, we're not using any `zipGlobPattern` or `zipIgnorePattern`, but we are using `files.exclude` in `.vscode/settings.json` to hide both the build folder and any .js files.

**Comment by nturinski** [comment]
This was some good insight.

> we are using files.exclude in .vscode/settings.json to hide both the build folder and any .js files.

From vscode API documentation:

* @param exclude  A glob pattern that defines files and folders to exclude. The glob pattern
		 * will be matched against the file paths of resulting matches relative to their workspace. When `undefined`, default file-excludes (e.g. the `files.exclude`-setting
		 * but not `search.exclude`) will apply. When `null`, no excludes will apply.

Since the project has no ignore array, I'm assuming that this `{${ignore.join(',')}}` ends up being a `''` which then gets converted to undefined, which then automatically uses the `files.exclude` setting.

Seems like the fix for us is to pass in `null` when we do not detect any `funcignore` file (or an empty one).

Sorry for the inconvenience, but the workaround right now seems like is to remove that `files.exclude` for the `.js` files.

**Comment by nturinski** [comment]
> though the test coverage could probably be improved 😅

sigh definitely agree with you there. It's part of our tech debt, but feels like we can't pay off the interest fast enough sometimes.

**Comment by nturinski** [comment]
@kristoffer-zliide Totally understandable.  We are planning another release soon.  

Would it be possible for in the meantime for you to have a `.funcignore` file wherever the deploy subpath is? It doesn't need anything meaningful, just some value that wouldn't return falsey/undefined and thus invoke VS Code's default behavior of using `files.exclude` setting when undefined is passed in.

**Comment by kristoffer-zliide** [comment]
Sorry, don't really see the difference between adding a dummy file or making a dummy modification to a file.

**Comment by nturinski** [comment]
@kristoffer-zliide Yeah, I see your point. Our current release plan is to ship the extension end of next week, but I can build you a custom vsix with the fix that you can install manually.

Let me know if you'd prefer the custom vsix solution.
